rust/projects/file-io/src/kv.rs

```diff
@@ -63,31 +59,137 @@ impl KvStore {
     ///
     /// # Error
     ///
-    /// It propagates I/O or serialization errors during writing the WAL.
+    /// It propagates I/O or serialization errors during writing the log.
     pub fn set(&mut self, key: String, value: String) -> Result<()> {
-        let cmd = SetCommand::new(key, value);
-        serde_json::to_writer(&mut self.wal_writer, &cmd)?;
-        self.wal_writer.flush()?;
-        self.map.insert(cmd.key, cmd.value);
+        self.kv_log.set(key, value)?;
+        if self.kv_log.uncompacted > COMPACTION_THRESHOLD {
+            self.compact()?;
+        }
         Ok(())
     }
 
     /// Gets the string value of a given string key.
     ///
     /// Returns `None` if the given key does not exist.
-    pub fn get(&self, key: String) -> Result<Option<String>> {
-        Ok(self.map.get(&key).cloned())
+    pub fn get(&mut self, key: String) -> Result<Option<String>> {
+        self.kv_log.get(key)
+    }
+
+    fn compact(&mut self) -> Result<()> {
+        // The new log file for merged entries
+        let tmp_log_path = self.path.join("data.log.new");
+        let mut new_writer = BufWriter::new(
+            OpenOptions::new()
+                .create(true)
+                .truncate(true)
+                .write(true)
+                .open(&tmp_log_path)?,
+        );
+
+        // Copy content in order to reduce seeks
+        let mut pos_vec: Vec<_> = self.kv_log.index.iter().collect();
+        pos_vec.sort_unstable_by_key(|(_, cmd_pos)| cmd_pos.pos);
+        let mut new_pos = 0; // pos in the new log file
+        let mut new_index = HashMap::new(); // index map for the new log file
+        for (key, cmd_pos) in pos_vec {
+            if self.kv_log.reader.pos != cmd_pos.pos {
+                self.kv_log.reader.seek(SeekFrom::Start(cmd_pos.pos))?;
+            }
+
+            let mut entry_reader = (&mut self.kv_log.reader).take(cmd_pos.len);
+            let len = io::copy(&mut entry_reader, &mut new_writer)?;
+            new_index.insert(key.clone(), (new_pos..new_pos + len).into());
+            new_pos += len;
+        }
+        drop(new_writer);
+
+        // As all entries are written to the log, we can safely rename it to a valid log file name
+        let log_path = new_log_path(&self.path);
+        fs::rename(tmp_log_path, &log_path)?;
+
+        // Reopen using the new file name
+        let mut kv_log = KvLog::open(&log_path)?;
+        // Use the index map built on writing instead of reloading the log file
+        kv_log.index = new_index;
+        // Update the KvLog we are using
+        mem::swap(&mut self.kv_log, &mut kv_log);
+
+        // Close old log file before removing it. (It's a must on Windows I think)
```

Comment:
It is generally, though I think there are options that can be passed to the win32 api that allow, e.g. delete-on-close.

Reply:
I thought the course focused on Rust, so I tried to avoid using any platform-specific approach. I was also avoiding use libc or winapi in this course as unsafe must be used. What's your opinion?

Reply:
Don't make any changes for this. I was just commenting on your comment.